reconcile/aws_ami_cleanup/integration.py

```diff
@@ -0,0 +1,262 @@
+) -> list[AWSAmi]:
+    """Get amis that match regex older than given age"""
+
+    images = aws_api.paginate(
+        ec2_client, "describe_images", "Images", {"Owners": [owner]}
+    )
+
+    pattern = re.compile(regex)
+    results = []
+    for i in images:
+        if not re.search(pattern, i["Name"]):
+            continue
+
+        creation_date = datetime.strptime(i["CreationDate"], "%Y-%m-%dT%H:%M:%S.%fZ")
+        current_delta = utc_now - creation_date
+        delete_delta = timedelta(seconds=age_in_seconds)
+
+        if current_delta < delete_delta:
+            continue
+
+        # We have nothing to do with untagged AMIs since we will need tags to verify if AMI are
+        # in use or not.
+        if not i.get("Tags"):
+            continue
+
+        tags = {AmiTag(**tag) for tag in i.get("Tags")}
+        results.append(
+            AWSAmi(image_id=i["ImageId"], tags=tags, creation_date=creation_date)
+        )
+
+    return results
+
+
+def get_region(
+    cleanup: Mapping[str, Any],
+    account: Mapping[str, Any],
+) -> str:
+    """Defines the region to search for AMIs."""
+    region = cleanup.get("region") or account["resourcesDefaultRegion"]
+    if region not in account["supportedDeploymentRegions"]:
+        raise ValueError(f"region {region} is not supported in {account['name']}")
+
+    return region
+
+
+def get_app_interface_amis(
+    namespaces: Optional[list[NamespaceV1]], ts: Terrascript
+) -> list[AIAmi]:
+    """Returns all the ami referenced in ASGs in app-interface."""
+    app_interface_amis = []
+    for n in namespaces or []:
+        for er in n.external_resources or []:
+            if not isinstance(er, NamespaceTerraformProviderResourceAWSV1):
+                continue
+
+            for r in er.resources:
+                if not isinstance(r, NamespaceTerraformResourceASGV1):
+                    continue
+
+                tags = set()
+                for i in r.image:
+                    if isinstance(i, ASGImageGitV1):
+                        tags.add(
+                            AmiTag(
+                                key=i.tag_name,
+                                value=ts.get_commit_sha(i.dict(by_alias=True)),
+                            )
+                        )
+                    elif isinstance(i, ASGImageStaticV1):
+                        tags.add(AmiTag(key=i.tag_name, value=i.value))
+
+                app_interface_amis.append(AIAmi(identifier=r.identifier, tags=tags))
+
+    return app_interface_amis
+
+
+def check_aws_ami_in_use(
+    aws_ami: AWSAmi, app_interface_amis: list[AIAmi]
+) -> Optional[str]:
+    """Verifies if the given AWS ami is in use in a defined app-interface ASG."""
+    for ai_ami in app_interface_amis:
+        # This can happen if the ASG init template has changed over the time. We don't have a way
+        # to properly delete these automatically since we cannot assure they are not in use.
+        if len(ai_ami.tags) > len(aws_ami.tags):
```

Comment:
Is `ai_ami.tags` defined in app interface? What if we add new tags but old AMI were built without them?